File: Tree/dfs/LeetCode/maximum-number-of-k-divisible-components.cpp
class Solution {
public:
    int cnt = 0;
    int maxKDivisibleComponents(int n, vector<vector<int>>& edges, vector<int>& values, int k) {
        vector<vector<int>> adj(n);
        for(auto &it: edges){
            adj[it[0]].push_back(it[1]);
            adj[it[1]].push_back(it[0]);
        }
        vector<int> vis(n, 0);
        vector<int> parent(n, 0);
        cnt = 0;
        int sum = 0;
        dfs(0, adj, values, k, vis, parent);
        return cnt;
    }
    void dfs(int u, vector<vector<int>>&adj, vector<int>&val, int k, vector<int>&vis, vector<int>&parent){
        vis[u] = true;
        for(int &v: adj[u]){
            if(vis[v] == false){
                parent[v] = u;
                dfs(v, adj, val, k, vis, parent);
            }
        }
        if(val[u] % k == 0){
            cnt++;
            val[u] = 0;
        }else{
           val[parent[u]] += val[u];
        }
    }
};

File: Graph/Topological-Sort/maximum-number-of-k-divisible-components.java
class Solution {
    public int maxKDivisibleComponents(int n, int[][] edges, int[] values, int k) {
        if (n < 2) {
            return 1;
        }

        int ans = 0;
        List<List<Integer>> adj = new ArrayList<>();
        int [] inDegree = new int[n];
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            adj.get(u).add(v);
            adj.get(v).add(u);
            inDegree[u]++;
            inDegree[v]++;
        }

        long[] longValues = new long[n];
        for (int i = 0; i < n; i++) {
            longValues[i] = values[i];
        }

        Queue<Integer> que = new LinkedList<>();
        for (int u = 0; u < n; u++) {
            if (inDegree[u] == 1) {
                que.offer(u);
            }
        }

        while (!que.isEmpty()) {
            int u = que.poll();
            inDegree[u]--;
            long val = 0;
            if (longValues[u] % k == 0) {
                ans++;
            } else {
                val = longValues[u];
            }

            for (int v : adj.get(u)) {
                if (inDegree[v] == 0) {
                    continue;
                }

                inDegree[v]--;
                longValues[v] += val;

                if (inDegree[v] == 1) {
                    que.offer(v);
                }
            }
        }

        return ans;
    }
}